### PR TITLE
[FIX] l10n installl time and determinism

### PR DESCRIPTION
**Pull request by Xavier-Do**
TLDR speedup l10n_standalone install  from  30-70+ minutes to 15-25 minutes Also speedup tests a little.

Sample build without much load (started in the morning)
https://runbot.odoo.com/runbot/batch/1417732/build/61186284
17 + 44
Sample build with worst case scenario simulation (no vacuum, 8 parallel build)
https://runbot.odoo.com/runbot/build/61186224
20 + 59

Still need to investigate a little on test_l10n part, to see potential effects and improvements.

The current l10n staging build is slow and wih a noticeable randomness, mainly for the install part. The install can take between 35 and 70+ minutes

# Investigation

To get some base info on the cause of this issue, a test was made on a reserved runbot to remove possible cause of randomness caused by other builds (there is always noise in test install time since the postgresql server is shared among builds, and a build could take more than one cpu when using external code (chrome, sphynx, ...)

When running [4 builds in parallel](https://runbot.odoo.com/runbot/batch/1406680/build/60827038), The time was around 32 minutes (quite consistent)

When running [8 builds in parallel](https://runbot.odoo.com/runbot/batch/1406739/build/60833055) the time goes up to around 42 minutes 

In this case the server load remains quite low but the postgresql server is quite busy, but not enough to disturb the garbage collection to much. 

For the next tests, the autovacuum was disabled on the server.
First, [with 4 builds parallel builds](https://runbot.odoo.com/runbot/batch/1407430/build/60842940), the average time ramps up to 45 minutes.
Second, with [8 parallel](https://runbot.odoo.com/runbot/batch/1407575/build/60845564) (more than that but the first 8 where in parallel) the time ramps up to ~70 minutes, the worst case scenario.

Some [further experiments ](https://runbot.odoo.com/runbot/build/61024217) were made making dummy analyze, making the build time more consistent (still high, mainly because of the many analyze) showing that even with this extra ANALYZE time the build was faster in average. 

Finally, to identify the exact cause, a profiling was made locally (should have been the first step maybe...) from the whole install, module by module to spot slow modules and slowness cause. 

```patch
diff --git a/odoo/modules/loading.py b/odoo/modules/loading.py
index 13845382e454..6928f8d04bc8 100644
--- a/odoo/modules/loading.py
+++ b/odoo/modules/loading.py
@@ -154,6 +154,7 @@ def load_module_graph(env, graph, status=None, perform_checks=True,
     import datetime
     prid = datetime.datetime.now().strftime('%Y%m%d%H%M%S')
     for index, package in enumerate(graph, 1):
+      with odoo.tools.profiler.Profiler(profile_session=env.cr.dbname + prid, description=package.name, db=env.cr.dbname):
         module_name = package.name
         module_id = package.id

```

And also on try_loading at the end

```patch
diff --git a/addons/account/tests/test_account_all_l10n.py b/addons/account/tests/test_account_all_l10n.py
index 9d4195cec61b..422d66c362fc 100644
--- a/addons/account/tests/test_account_all_l10n.py
+++ b/addons/account/tests/test_account_all_l10n.py
@@ -73,7 +73,9 @@ def test_all_l10n(env):
         env.user.company_id = company
         _logger.info('Testing COA: %s (company: %s)', template_code, company.name)
         try:
-            env['account.chart.template'].with_context(l10n_check_fields_complete=True).try_loading(template_code, company, install_demo=True)
+            from odoo.tools import profiler
+            with profiler.Profiler(db=env.cr.dbname, collectors=[profiler.PeriodicCollector(interval=0.1)]):
+                env['account.chart.template'].with_context(l10n_check_fields_complete=True).try_loading(template_code, company, install_demo=True)
             env.cr.commit()
         except Exception:
             _logger.error("Error when creating COA %s", template_code, exc_info=True)

```

## _adapt_parent_account_group

This reveled the main cause of the slowness,  `_adapt_parent_account_group`
Slow, even without randomness
image
image
One example of when the randomness strikes (query took 1 minutes 45 second)
image

[Speedscope.zip](https://github.com/odoo/odoo/files/14928528/Speedscope.zip)

The current behavior will actually update parent_path for **all** account_group in the database after each try_loading, (and create or update of an account_group when postpone is not enabled), even if no modification on group were made. For some of them the query is slow because table statistics are broken by many updates, and the list of id is important, leading to sub-optimal query plan when the database auto-vacuum is to busy to analyze the table in time. (some queries taking around 2 minutes minute) This is also caused by an issue solved in the second commit, where all account_group are checked while it is intended ton only check those of the current company.

This pr tries to improve this on targeting records that have change. The [first result ](https://runbot.odoo.com/runbot/batch/1413399/build/61033322 ) looks like we can reach ~30 minutes per build, at full load, even with auto_vacuum disables (a manual analyze was added on the table if a lot of operations were made) (Note that some build are failling randomly because of random error logged in profiler frames, can be ignored)

On a normal build, without artifical load and disabled auto-vacuum, the install time is around 15-25 minutes 

Still need to reconsider why all of this is made by sql, and if it could be done another way, this could also be done using the orm computed fields for a code easier to read. The main issue is to have a good selection of record to update.

After this first change, some other aspect were checked, and it looks like the build may have become slower after #156238@
This was suggested by @gawa-odoo and checked by @ryv-odoo, this is True but it was discovered that the previous usage was wrong and that parent_path in stable are not correct. The conclusion was made that to solve all issue, we can just remove the parent_store from the model and stop updating the parent_path.

## commit instead of savepoint

This change was not tested independently, but before the combined fix, generating the coa could take between 
7 and 20+ minutes. Keeping a transaction opened for too long can increase the load on the server in some cases, and it looks like the goal is mainly to be able to install other coa even if one fail at the beginning. Replacing by a commit between each coa should do the trick (only in test_all_l10n, so shouldn't arm any real usage). 

## _adapt_accounts_for_account_groups

image
_adapt_accounts_for_account_groups is also slow in some case, because the queries is made with a select containing a where close containing all id of the database, making a suboptimal query plan.
A second commit tries to fix that by filtering on the company instead to change the behavior. 
This can be more than 1 second per module for the last one. 
It could maybe remove 1~2 minutes on the build time, no real performance test was made to check that. 
The previous behavior was to select all record, targeting the company one but this is not the case when executed in sudo.
This change proposal should increases readability, avoid mistakes and improve performances.

### CODE REVIEW

File: addons/account/models/chart_template.py [hunk 1]
```diff
@@ -211,8 +211,10 @@ def _load(self, template_code, company, install_demo):
 
         # Manual sync because disable above (delay_account_group_sync)
         AccountGroup = self.env['account.group'].with_context(delay_account_group_sync=False)
-        AccountGroup._adapt_accounts_for_account_groups(self.env['account.account'].search([]))
-        AccountGroup.search([])._adapt_parent_account_group()
+        AccountGroup._adapt_accounts_for_account_groups(check_all=True)
+        AccountGroup._adapt_parent_account_group(check_all=True)
```
Comment (qdp-odoo) [inline addons/account/models/chart_template.py:215]: I don't think there exists a use case where we want to adapt accounts from other companies (except for branches setup)

so a more elegant solution would be to pass the company as parameter to _adapt_accounts_for_account_group() and _adapt_parent_account_group() so that the subquery is efficient.

```suggestion
        AccountGroup._adapt_accounts_for_account_groups(company=company.id)
        AccountGroup._adapt_parent_account_group(company.id)
```

**Review by qdp-odoo — COMMENTED** [review]
Thanks for the detailed explanation and the thorough analysis :+1: despite the PR still being a WIP :)

I don't see anything obviously wrong here, the old logic seems maintained and nothing should be broken...

Yet, I agree that the fix with check_all is a bit awkward, I'd advocate testing and benchmarking with a company ID given as parameter, as I don't think it's ever relevant to recompute account groups for ALL account: we're always supposed to be able to filter on companies

File: addons/account/models/chart_template.py (hunk 1 above)
Comment (Xavier-Do) [reply addons/account/models/chart_template.py:215]: I will check, I'd like to be even more precise by selecting the record that were updates. It looks like an issue during tests, the setupclass is also doing a tryloading when setuping companies and it is slow.... but your solution could actually help there. Will make some experiment on another branch and keep you in touch once I have result, thanks for the review!

File: addons/account/models/account_account.py [hunk 2]
```diff
@@ -1038,12 +1043,21 @@ def _adapt_parent_account_group(self):
             UPDATE account_group child
                SET parent_id = relation.parent_id
               FROM relation
-             WHERE child.id = relation.child_id;
+             WHERE child.id = relation.child_id
+               AND (child.parent_id != relation.parent_id OR (child.parent_id IS NULL AND relation.parent_id IS NOT NULL)) -- avoid to update if nothing changed
```
Comment (Xavier-Do) [inline addons/account/models/account_account.py:1043]: is not distinct from

File: addons/account/models/account_account.py [hunk 3]
```diff
@@ -1017,8 +1020,14 @@ def _adapt_parent_account_group(self):
         """
         if self.env.context.get('delay_account_group_sync'):
             return
-        if not self:
+        if company:
+            company_ids = company.ids
+        else:
+            company_ids = self.company_id.ids
```
Comment (william-andre) [inline addons/account/models/account_account.py:1026]:
```suggestion
        company_ids = (company or self.company_id).ids
```

File: addons/account/models/account_account.py [hunk 4]
```diff
@@ -1017,8 +1020,14 @@ def _adapt_parent_account_group(self):
...
+
+        if not company_ids:
             return
+        cond = SQL('child.company_id IN %s', tuple(company_ids))
```
Comment (william-andre) [inline addons/account/models/account_account.py:1030]: Why is this outside of the main query? :thinking:

File: addons/account/models/account_account.py [hunk 5]
```diff
@@ -1032,18 +1041,22 @@ def _adapt_parent_account_group(self):
...
         """
-        self.env.cr.execute(query, {'company_ids': tuple(self.company_id.ids)})
-        self.invalidate_model(['parent_id'])
-        self.search([('company_id', 'in', self.company_id.ids)])._parent_store_update()
+        self.env.cr.execute(SQL(query, cond))
 
+        updated_record = self.env.cr.fetchall()
+        if updated_record:
+            self.invalidate_model(['parent_id'])
+            # self.env.cr.execute('ANALYZE account_group') to check, could be usefull on some cases if number of update was high
```
Comment (william-andre) [inline addons/account/models/account_account.py:1059]: :see_no_evil:

File: addons/account/tests/test_account_all_l10n.py [hunk 6]
```diff
@@ -60,14 +62,19 @@ def try_loading_patch(self, template_code, company, install_demo=True):
         }
         for template_code, template in not_loaded_codes
     ])
-
+    env.cr.commit()
     # Install the CoAs
+    start = time.time()
+    env.cr.execute('ANALYZE')
+    logger = logging.getLogger('odoo.loading')
+    logger.runbot('ANALYZE took %s seconds', time.time() - start)  # not sure this one is usefull
```
Comment (william-andre) [inline addons/account/tests/test_account_all_l10n.py:70]: :see_no_evil:

File: addons/account/models/account_account.py [hunk 7]
```diff
@@ -859,13 +862,12 @@ def _merge_method(self, destination, source):
...
     _name = "account.group"
     _description = 'Account Group'
-    _parent_store = True
     _order = 'code_prefix_start'
     _check_company_auto = True
     _check_company_domain = models.check_company_domain_parent_of
 
     parent_id = fields.Many2one('account.group', index=True, ondelete='cascade', readonly=True, check_company=True)
-    parent_path = fields.Char(index=True)
+    parent_path = fields.Char(index=True)  # unused, can be removed
```
Comment (william-andre) [inline addons/account/models/account_account.py:870]: Since we are in master, we can do it

**Review by william-andre — COMMENTED** [review]
Looking good overall :+1: 

Why are we only targetting `master` though?
Aren't most of the changes relevant in earlier versions too?

File: addons/account/models/account_account.py [hunk 8]
```diff
@@ -859,13 +862,12 @@ def _merge_method(self, destination, source):
 class AccountGroup(models.Model):
     _name = "account.group"
     _description = 'Account Group'
-    _parent_store = True
```
Comment (rco-odoo) [inline addons/account/models/account_account.py:862]: Having `_parent_store = True` comes with its own mechanism to guarantee the absence of cycles.  Removing the `_parent_store` implies to re-add an explicit constraint for that.

**Comment by Xavier-Do** [comment]
> Why are we only targetting master though?
> Aren't most of the changes relevant in earlier versions too?

Master is the branch that suffer the most from the issue and I prefer to merge this kind of change in master first to test and validate, and then backport. It is intended to backport this at some point for sure.

File: addons/account/models/account_account.py (hunk 4 above)
Comment (Xavier-Do) [reply addons/account/models/account_account.py:1030]: Was conditional at some point

File: addons/account/models/account_account.py (hunk 7 above)
Comment (Xavier-Do) [reply addons/account/models/account_account.py:870]: I prefer to do that in another pr for a a faster merge and easier backport. (justified in commit message)

File: addons/account/models/account_account.py (hunk 8 above)
Comment (rco-odoo) [reply addons/account/models/account_account.py:862]: Oh damned, the constraint was actually present.  It was redundant, it now is useful.

**Review by rco-odoo — APPROVED** [review]
Looks good for me.

**Review by ryv-odoo — APPROVED** [review]

**Comment by rco-odoo** [comment]
@robodoo r+

**Comment by rco-odoo** [comment]
@robodoo rebase-ff

**Comment by rco-odoo** [comment]
Followup https://github.com/odoo/odoo/pull/161798.
